### PR TITLE
change pretty default output localTime

### PR DESCRIPTION
**Pull request by victor0801x**
change pretty.formatTime parse to localTime
(original is convert date.toISOString())

note: this is my first time pr, if do any wrong, I'm very sorry

### CODE REVIEW

**Comment by mcollina** [comment]
Thanks for your contribution! The build on travis is failing, you might have changed some of the existing functionality. Can you check? you can run the unit tests via npm test. 

Can you also add some to cover the new functionality?

**Comment by victor0801x** [comment]
this should not break anything on tests. 
```
test/addlevel.test.js ............................... 21/21
test/basic.test.js .................................. 98/98
test/browser.test.js .............................. 168/168
test/error.test.js .................................. 10/10
test/escaping.test.js ............................... 39/39
test/events.test.js ................................. 13/13 6s
test/extreme.test.js .................................. 7/7 2s
test/http.test.js ................................... 19/19
test/levels.test.js ................................. 37/37
test/metadata.test.js ............................... 24/24
test/pretty.test.js ................................. 43/43
test/serializers.test.js .............................. 9/9
test/timestamp.test.js .............................. 16/16
total ............................................. 504/504

  504 passing (13s)

  ok

```
only changes pretty.js internal `formatTime()` to call `asLocalDate()`, return local-time, `'YYYY-MM-DDThh:mm:ss.SSSTZ', // '2017-08-05T12:30:45.789+08:00'`, instead of `asISODate()`, return utc iso-format-time.

currently, does not support custom input format time string. I need to figure out how to parse bin.js command input custom string, then pass to pretty.js option

**Comment by mcollina** [comment]
I think adding unit tests to this PR should be a first priority. You can use http://npm.im/minimist to parse input, if we aren't already doing so.

**Comment by victor0801x** [comment]
sorry, I'm not familiar with github. i close first

**Comment by mcollina** [comment]
Why did you close? It's ok ;). I'm happy to review things as you upload them.

**Comment by jsumners** [comment]
@victor0801x sorry about this, but I merged in a couple other PRs that cause this to have conflicts. Will you resolve them so this can be merged?

**Comment by jsumners** [comment]
#317 also provides an improved API for the custom formatter function. You may want to investigate using it to solve this issue.

**Comment by victor0801x** [comment]
sorry for reply so late, recently do not have spare time til March, hope someone can improve or fix this

**Comment by victor0801x** [comment]
@jsumners I fixed the conflict, but ci is failed ?

**Comment by jsumners** [comment]
This test is failing:

https://github.com/pinojs/pino/blob/c77d8ec5ce37f56cbdf1689d4ec9ac585eb6b74b/test/browser.test.js#L217-L230

@davidmarkclements any ideas why?

@victor0801x you have left some dead code and a TODO in place:

https://github.com/victor0801x/pino/blob/c37a2acc5d621bd96f412ac007cffc420b694e4b/pretty.js#L189-L195

Why?

**Comment by victor0801x** [comment]
1. original `blabla |pino -t ` will convert Epoch timestamp into iso timestamp by call asISODate (UTC)
This pr will convert into local-iso-timestamp by calling asLocalDate
Not sure, everybody accept. also, not sure asISODate  function should be remove or not ? 
so I temporarily comment in order to pass the test.

2. I think maybe can convert into user specified timezone by ` |pino -t +8` or ` |pino --tz +8`.
also not sure this is necessary when default is convert into localISODate

**Comment by jsumners** [comment]
My suggestion in https://github.com/pinojs/pino/pull/280#issuecomment-342483584 was to consider providing an alternative formatter function that outputs the date in this manner. That would make it easier to swap between the two types of logging.

**Comment by victor0801x** [comment]
how can I pass the formater function from cli ?

**Comment by jsumners** [comment]
As with your current approach, which is fine I think, you'd have to add some sort of switch to toggle them.

**Comment by davidmarkclements** [comment]
I think it may be due to a patch update to quick-format-unescaped - there was an escaping bug for a particular case - need to look into it more

**Comment by jsumners** [comment]
@pinojs/core what do you think of this PR?

File: test/pretty.test.js [hunk 1]
```diff
@@ -367,6 +367,24 @@ test('works without time', function (t) {
...
+    t.ok(line.match(/(?!^)INFO.*/), 'includes level')
+    t.ok(line.indexOf('' + process.pid) > 0, 'includes pid')
+    t.ok(line.indexOf('' + hostname) > 0, 'includes hostname')
+    t.ok(Date.parse(line.substr(1, 29)) === parseInt(now), 'local date timezone')
+    return line
+  }))
+
+  var instance = pino(prettier)
+
+  instance.info('>' + Date.now() + '<' + ' hello world')
```
Comment (mcollina) [inline test/pretty.test.js]: This Date.now() may not be doing what you think. Can you remove it?

File: pretty.js [hunk 2]
```diff
@@ -23,6 +23,27 @@ var standardKeys = [
   'v'
 ]
 
+function toTimezoneOffset (aMinTimeoffset) {
+  // +/- minute timeoffset
+  var tz = aMinTimeoffset || new Date().getTimezoneOffset()
+  var tmp = Math.abs(tz)
+
+  // var offset = String(Math.floor(tmp / 60)).padStart(2, '0') + ':' + String(tmp % 60).padStart(2, '0')
```
Comment (mcollina) [inline pretty.js]: can you remove this commented out line?

File: pretty.js [hunk 3]
```diff
@@ -155,17 +177,43 @@ function pretty (opts) {
...
+    var _format = format
+      .replace(/Y{1,4}/g, year)
+      .replace(/MM/g, month)
+      .replace(/DD/g, day)
+      .replace(/hh/g, hour)
+      .replace(/mm/g, minute)
+      .replace(/ss/g, second)
+      .replace(/SSS/g, milli)
+      .replace(/TZ/g, toTimezoneOffset(tzOffset))
+    return _format
```
Comment (mcollina) [inline pretty.js]: are we sure there is no common utility we can use instead of this block of replace?

File: test/pretty.test.js (hunk 1 above)
Comment (victor0801x) [reply test/pretty.test.js]:
````js
instance.info('>' + Date.now() + '<' + ' hello world')
var now = line.slice(line.indexOf('>') + 1, line.indexOf('<'))
t.ok(Date.parse(line.substr(1, 29)) === parseInt(now), 'local date timezone')
````
According to MDN, `Date.now()` will return the number of milliseconds elapsed since January 1, 1970 00:00:00 UTC. right?
this is what I try to do
```
[2018-02-10T17:32:02.719+08:00] INFO (8484 on jimmy): >1518255122719< hello world
Date.parse('2018-02-10T17:32:02.719+08:00') === 1518255122719
           ^^^ local datetime with tz ^^^
```

File: pretty.js (hunk 3 above)
Comment (victor0801x) [reply pretty.js]: I think this is minimal implemantation without any deps.
this is a log libary not a datetime libary !? XD
there must have another fast convert implemantation

File: test/pretty.test.js (hunk 1 above)
Comment (mcollina) [reply test/pretty.test.js]: I don't think this is a good way to test this. You are assuming that the time picked from `instance.info()` is the same of the log message. However, this is not guaranteed. 
A better approach might be to mock `Date.now()`.

File: pretty.js (hunk 3 above)
Comment (mcollina) [reply pretty.js]: Can you make the _previous code_ as the default (`new Date(time).toISOString()`)?

Comment (victor0801x) [reply pretty.js]: sure, add new function asLocalISODate and keep asISODate as the previouse code
and add CLI options (`--localTime`) transform to localISOTime with tz(`YYYY-MM-DDThh:mm:ss.SSSTZ`)
but, not sure how to add unit test for `asLocalISODate`, only bin.js for pretty print

File: test/pretty.test.js [hunk 4]
```diff
@@ -354,7 +354,7 @@ test('works without time', function (t) {
   var prettier = pretty()
   prettier.pipe(split(function (line) {
     t.ok(line.match(/.*hello world$/), 'end of line matches')
-    t.ok(line.match(/^INFO.*/), 'includes level')
+    t.ok(line.match(/(?!^)INFO.*/), 'includes level')
```
Comment (mcollina) [inline test/pretty.test.js]: why this change?

Comment (victor0801x) [reply test/pretty.test.js]: my bad, I finally figure out why previouse code doing this
```js
    line = (levelFirst)
        ? asColoredLevel(value) + ' ' + formatTime(value)
        : formatTime(value, ' ') + asColoredLevel(value)
```
I change levelFirst code  into
```js
    line = (levelFirst)
        ? asColoredLevel(value) + ' ' + formatTime(value)
        : formatTime(value) + ' ' + asColoredLevel(value)
```
cause the start of output is `a space + level`, then missing time test will fail
i will make this to previouse code

Comment (victor0801x) [reply test/pretty.test.js]: I remove the previouse test I add, because the default output is `asISODate`, I do not have to add this test again. 
I need some help, how to add test for `asLocalISODate`, it only expose to CLI use. Separate code into js? expose method from pretty.js?

Comment (jsumners) [reply test/pretty.test.js]: Simplest (being a relative term) would be to use `child_process.fork` and parse the `stdout` stream of the child.

**Comment by victor0801x** [comment]
not sure, why node v6.x ci failed !!
I add --localTime transform to local-iso-date & --dateFormat for parse custom format test
@mcollina @jsumners any advises

**Comment by jsumners** [comment]
No idea. I restarted the job and it worked.

File: test/pretty.test.js [hunk 5]
```diff
@@ -380,3 +381,66 @@ test('does not throw error when enabled with stream specified', function (t) {
...
+
+  child.on('close', function () {
+    line = line.slice(0, -1 * (os.EOL).length)
+    var formatDate = line.slice(line.indexOf('[') + 1, line.indexOf(']'))
+    var toISODate = formatDate.replace(/\//g, '-').replace(/,/g, ':').replace(/_/g, '.')
+    t.ok(line.match(/.*hello world$/), 'end of line matches')
+    t.ok(line.match(/(?!^)INFO.*/), 'includes level')
+    t.ok(line.indexOf('' + process.pid) > 0, 'includes pid')
+    t.ok(line.indexOf('' + hostname) > 0, 'includes hostname')
+    t.ok(Date.parse(toISODate) === parseInt(now), 'custDateFormat <-> UTC match')
```
Comment (mcollina) [inline test/pretty.test.js]: This needs to be checked using an interval. These numbers are not assured to be identical on a CI system, as they are dependent on what else is running.

Comment (victor0801x) [reply test/pretty.test.js]: how can i check custom dateformat or local-time without passing epo

Comment (victor0801x) [reply test/pretty.test.js]: how can I check transformed dateformat or local-time is correct? 
this is my thought, I pass time from msg (log.time = log.msg), then i can check the datestring

Comment (mcollina) [reply test/pretty.test.js]: How about 

```
t.ok(Date.parse(toISODate) > parseInt(now) - 2000)
t.ok(Date.parse(toISODate) < parseInt(now) + 2000)
```

Comment (victor0801x) [reply test/pretty.test.js]: sure, I misunderstand your words. 
by the way, is there any suggestions/tricks for comparing floating-point ?
